webpage customization support added: annotationDiv customizationHook

paul-shannon opened the pull request:
@davidjamesbeck Two small changes:

  ijalLine.py now inserts, after every translation line, an empty 

```<div class="annotationDiv"></div>```.

  text.py adds a comment line to the header 

```
  <head>
    <meta charset="UTF-8"/>
    <script src="jquery-3.3.1.min.js"></script>
    <link rel = "stylesheet" type = "text/css" href = "ijal.css" />
    <!-- customizationHook -->
  </head>
```
I replace this comment, after building a page, using the venerable unix tool ```sed```.  In my current work, this one line is replaced by these includes, which support the popup annotations:
```
<link rel="stylesheet" type="text/css" href="../pshannon.css" />
<script src="../showdown.js"></script>
<script src="../pshannon.js"></script>
<script src="../kb.js"></script>
```

paul-shannon commented:
@davidjamesbeck  - Hi David,

Do you wish to review and merge this PR?  It should slide right in, having no conflicts.
 
 - Paul

davidjamesbeck commented:
Done. I’d run the tests and looked at the output days ago, but didn’t see it as a separate branch in my local repo for some reason.

Have you checked out webapp2_run_locally.py (dumb name, I know)? If I get a chance to make a new video tutorial soon, I’ll probably rechristen it webapp3.py and try it out on our VM

David

David

> On Nov 26, 2019, at 6:28 PM, Paul Shannon <notifications@github.com> wrote:
> 
> @davidjamesbeck <https://github.com/davidjamesbeck> - Hi David,
> 
> Do you wish to review and merge this PR? It should slide right in, having no conflicts.
> 
> Paul
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/paul-shannon/slexil/pull/53?email_source=notifications&email_token=AKN4HBQHVP7Q4NRJXYRHXG3QVXEKZA5CNFSM4JQRGKBKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFH7DIY#issuecomment-558887331>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKN4HBVPNVYJ62EW45ZEB23QVXEKZANCNFSM4JQRGKBA>.
>

paul-shannon commented:
@davidjamesbeck  Hi David,

Many thanks for the merge of the PR.

The new drag-and-drop is really very fine!  I   Good progress and error messages, and good resilience in the face of errors.  I dragged the Inferno sound file into the eaf dialog, my mistake was handled smoothly.   I know how much work lies behind smooth UI operations like this.  Bravo!

Maybe ijal.css is, in my copy of the repo, in the wrong place?   Here’s my rendered page:

![image](https://user-images.githubusercontent.com/2480712/69736168-d62cbc80-10e6-11ea-95e8-9cfdf3c26e1f.png)

davidjamesbeck commented:
Interesting. I didn’t have that happen on my computer, but the HTML in the iFrame should be calling the CSS packaged with the HTML and other output in the PROJECTS folder, and the app is supposed to copy ijal.css to there, so the setup in your repo shouldn’t matter. If you don’t get a copy in that directory, maybe there’s an order-of-operations thing. I’ll run it again tonight if I get a chance and see if I can replicate the error.

Another possibility is that cacheing problem we were running into way back when. I spent 3 extra hours testing this last update because the iFrame kept showing me only 2 lines of Inferno. It turned out that, because I used 2 lines only to save time while fixing up the drag-and-drop, that version of the text was in a cache somewhere and got used instead of the 4 line version I wanted to see for the final tests. That’s one reason the new code is so much cleaner—I went through line by line looking for a failure point, and was cleaning up as I went. Anyway, maybe if you empty the cache, quit the browser, and do it over, it will work?

David

> On Nov 27, 2019, at 8:25 AM, Paul Shannon <notifications@github.com> wrote:
> 
> @davidjamesbeck Hi David,
> 
> Many thanks for the merge of the PR.
> 
> The new drag-and-drop is really very fine! I Good progress and error messages, and good resilience in the face of errors. I dragged the Inferno sound file into the eaf dialog, my mistake was handled smoothly. I know how much work lies behind smooth UI operations like this. Bravo!
> 
> Maybe ijal.css is, in my copy of the repo, in the wrong place? Here’s my rendered page:
> 
> 
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

davidjamesbeck commented:
I just cloned the GitHub repo to my office computer and I’m getting the same thing. The ijal.css file is in the right place and the HTML displays correctly if I open the Inferno.html file in a browser directly, so the iFrame seems not to be using our stylesheet.  I wonder if adding the ijal.css styles to the styles for the web app would have any effect?

I really don’t like dcc.iFrame, not at all. :-(

> On Nov 27, 2019, at 8:47 AM, David J Beck <dbeck@ualberta.ca> wrote:
> 
> Interesting. I didn’t have that happen on my computer, but the HTML in the iFrame should be calling the CSS packaged with the HTML and other output in the PROJECTS folder, and the app is supposed to copy ijal.css to there, so the setup in your repo shouldn’t matter. If you don’t get a copy in that directory, maybe there’s an order-of-operations thing. I’ll run it again tonight if I get a chance and see if I can replicate the error.
> 
> Another possibility is that cacheing problem we were running into way back when. I spent 3 extra hours testing this last update because the iFrame kept showing me only 2 lines of Inferno. It turned out that, because I used 2 lines only to save time while fixing up the drag-and-drop, that version of the text was in a cache somewhere and got used instead of the 4 line version I wanted to see for the final tests. That’s one reason the new code is so much cleaner—I went through line by line looking for a failure point, and was cleaning up as I went. Anyway, maybe if you empty the cache, quit the browser, and do it over, it will work?
> 
> David
> 
>> On Nov 27, 2019, at 8:25 AM, Paul Shannon <notifications@github.com> wrote:
>> 
>> @davidjamesbeck Hi David,
>> 
>> Many thanks for the merge of the PR.
>> 
>> The new drag-and-drop is really very fine! I Good progress and error messages, and good resilience in the face of errors. I dragged the Inferno sound file into the eaf dialog, my mistake was handled smoothly. I know how much work lies behind smooth UI operations like this. Bravo!
>> 
>> Maybe ijal.css is, in my copy of the repo, in the wrong place? Here’s my rendered page:
>> 
>> 
>> 
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub, or unsubscribe.
>> 
>

paul-shannon commented:
> 
> I really don’t like dcc.iFrame, not at all. :-(

When using someone else’s library means we consent to their ontology, to their metaphysics: what the entities are, what operations are possible.   Sometimes those notions are - to use the old hacker term - manifestly brain dead.  Or at least that’s how it feels when things don’t work, make no sense, but (by our lights) ought to be straightforward.   “underdeveloped” is a kinder term.   Sorry for the pain!

 - Paul

davidjamesbeck commented:
Don’t be sorry, on the whole it is been pleasant and vastly educational. For all its quirks (which I am only sophisticated enough to perceive as “X is annoying for reasons I feel are not down to my incompetence”), Dash has given us the results we need with relatively little pain. I am actually hopeful it will develop, there seem to be users and people actively troubleshooting it. Let’s hope that we are only experiencing the pain of the early adopter (the “bleeding edge,” as a friend of mine put it).

> On Nov 27, 2019, at 12:21 PM, Paul Shannon <notifications@github.com> wrote:
> 
> 
> > 
> > I really don’t like dcc.iFrame, not at all. :-(
> 
> When using someone else’s library means we consent to their ontology, to their metaphysics: what the entities are, what operations are possible. Sometimes those notions are - to use the old hacker term - manifestly brain dead. Or at least that’s how it feels when things don’t work, make no sense, but (by our lights) ought to be straightforward. “underdeveloped” is a kinder term. Sorry for the pain!
> 
> - Paul
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/paul-shannon/slexil/pull/53?email_source=notifications&email_token=AKN4HBWRXL5GOGFLL6XDPL3QV3CEVA5CNFSM4JQRGKBKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFKQLRQ#issuecomment-559220166>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKN4HBVDTP5L6SLLRLARI6DQV3CEVANCNFSM4JQRGKBA>.
>

davidjamesbeck commented:
I think I’ve fixed this: I had altered the server route to only include “.html” files because I’d noticed the server route callback triggering multiple times, once for every file in the PROJECTS directory. I added “.css” to the accepted file suffixes and I see a properly formatted text in the iFrame on my home computer. You might want to pull the new webapp2_run_locally.py and try it out on yours as well. 

David

> On Nov 27, 2019, at 12:21 PM, Paul Shannon <notifications@github.com> wrote:
> 
> 
> > 
> > I really don’t like dcc.iFrame, not at all. :-(
> 
> When using someone else’s library means we consent to their ontology, to their metaphysics: what the entities are, what operations are possible. Sometimes those notions are - to use the old hacker term - manifestly brain dead. Or at least that’s how it feels when things don’t work, make no sense, but (by our lights) ought to be straightforward. “underdeveloped” is a kinder term. Sorry for the pain!
> 
> - Paul
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/paul-shannon/slexil/pull/53?email_source=notifications&email_token=AKN4HBWRXL5GOGFLL6XDPL3QV3CEVA5CNFSM4JQRGKBKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFKQLRQ#issuecomment-559220166>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKN4HBVDTP5L6SLLRLARI6DQV3CEVANCNFSM4JQRGKBA>.
>

paul-shannon commented:
The revisions to webapp2_run_locally.py work - the iframed page displays just right!

 - Paul

> On Nov 27, 2019, at 2:54 PM, David Beck <notifications@github.com> wrote:
> 
> I think I’ve fixed this: I had altered the server route to only include “.html” files because I’d noticed the server route callback triggering multiple times, once for every file in the PROJECTS directory. I added “.css” to the accepted file suffixes and I see a properly formatted text in the iFrame on my home computer. You might want to pull the new webapp2_run_locally.py and try it out on yours as well. 
> 
> David
> 
> > On Nov 27, 2019, at 12:21 PM, Paul Shannon <notifications@github.com> wrote:
> > 
> > 
> > > 
> > > I really don’t like dcc.iFrame, not at all. :-(
> > 
> > When using someone else’s library means we consent to their ontology, to their metaphysics: what the entities are, what operations are possible. Sometimes those notions are - to use the old hacker term - manifestly brain dead. Or at least that’s how it feels when things don’t work, make no sense, but (by our lights) ought to be straightforward. “underdeveloped” is a kinder term. Sorry for the pain!
> > 
> > - Paul
> > 
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub <https://github.com/paul-shannon/slexil/pull/53?email_source=notifications&email_token=AKN4HBWRXL5GOGFLL6XDPL3QV3CEVA5CNFSM4JQRGKBKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFKQLRQ#issuecomment-559220166>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKN4HBVDTP5L6SLLRLARI6DQV3CEVANCNFSM4JQRGKBA>.
> > 
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

davidjamesbeck commented:
Woohoo! 

Do you know what is going on with the server route callback exactly? Telling the webapp to fill the iFrame triggers that callback somehow, but I’m not sure why anything other than the HTML file is passed through it. Would it be because the HTML contains a link the ijal.css file?

David

> On Nov 27, 2019, at 7:08 PM, Paul Shannon <notifications@github.com> wrote:
> 
> The revisions to webapp2_run_locally.py work - the iframed page displays just right!
> 
> - Paul
> 
> > On Nov 27, 2019, at 2:54 PM, David Beck <notifications@github.com> wrote:
> > 
> > I think I’ve fixed this: I had altered the server route to only include “.html” files because I’d noticed the server route callback triggering multiple times, once for every file in the PROJECTS directory. I added “.css” to the accepted file suffixes and I see a properly formatted text in the iFrame on my home computer. You might want to pull the new webapp2_run_locally.py and try it out on yours as well. 
> > 
> > David
> > 
> > > On Nov 27, 2019, at 12:21 PM, Paul Shannon <notifications@github.com> wrote:
> > > 
> > > 
> > > > 
> > > > I really don’t like dcc.iFrame, not at all. :-(
> > > 
> > > When using someone else’s library means we consent to their ontology, to their metaphysics: what the entities are, what operations are possible. Sometimes those notions are - to use the old hacker term - manifestly brain dead. Or at least that’s how it feels when things don’t work, make no sense, but (by our lights) ought to be straightforward. “underdeveloped” is a kinder term. Sorry for the pain!
> > > 
> > > - Paul
> > > 
> > > —
> > > You are receiving this because you were mentioned.
> > > Reply to this email directly, view it on GitHub <https://github.com/paul-shannon/slexil/pull/53?email_source=notifications&email_token=AKN4HBWRXL5GOGFLL6XDPL3QV3CEVA5CNFSM4JQRGKBKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFKQLRQ#issuecomment-559220166>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKN4HBVDTP5L6SLLRLARI6DQV3CEVANCNFSM4JQRGKBA>.
> > > 
> > 
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub, or unsubscribe.
> > 
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/paul-shannon/slexil/pull/53?email_source=notifications&email_token=AKN4HBVRO3KCK4XVCW2F52LQV4RYZA5CNFSM4JQRGKBKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFLGZ7A#issuecomment-559312124>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKN4HBRO5LWP7ZJOE4RHBN3QV4RYZANCNFSM4JQRGKBA>.
>

paul-shannon commented:
>  would it be because the HTML contains a link the ijal.css file?

I think so.  That results in an http GET request which gets channeled through the specified flask route.   I think!

 - Paul

davidjamesbeck commented:
Ah. And the GET request is triggered when the iFrame tries to render the page?

> On Nov 28, 2019, at 9:36 AM, Paul Shannon <notifications@github.com> wrote:
> 
> 
> > would it be because the HTML contains a link the ijal.css file?
> 
> I think so. That results in an http GET request which gets channeled through the specified flask route. I think!
> 
> - Paul
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/paul-shannon/slexil/pull/53?email_source=notifications&email_token=AKN4HBSAKEBVOLTQ2T5L43LQV7XR5A5CNFSM4JQRGKBKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFNCGQY#issuecomment-559555395>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKN4HBQ6OVZK5GDTSHRC5Q3QV7XR5ANCNFSM4JQRGKBA>.
>

paul-shannon commented:
I think so.  I imagine there is a route invoked for the iFrame, and then subsequent routes (back to the flask http server) for every request embedded in the iFrame HTML which resolves to the flask http server.  



> On Nov 28, 2019, at 8:44 AM, David Beck <notifications@github.com> wrote:
> 
> Ah. And the GET request is triggered when the iFrame tries to render the page?
> 
> > On Nov 28, 2019, at 9:36 AM, Paul Shannon <notifications@github.com> wrote:
> > 
> > 
> > > would it be because the HTML contains a link the ijal.css file?
> > 
> > I think so. That results in an http GET request which gets channeled through the specified flask route. I think!
> > 
> > - Paul
> > 
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub <https://github.com/paul-shannon/slexil/pull/53?email_source=notifications&email_token=AKN4HBSAKEBVOLTQ2T5L43LQV7XR5A5CNFSM4JQRGKBKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFNCGQY#issuecomment-559555395>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKN4HBQ6OVZK5GDTSHRC5Q3QV7XR5ANCNFSM4JQRGKBA>.
> > 
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>
